packages/react-reconciler/src/ReactFiberClassComponent.js

```diff
@@ -620,6 +620,20 @@ export default function(
     instance.refs = emptyObject;
     instance.context = getMaskedContext(workInProgress, unmaskedContext);
 
+    warning(
+      instance.state !== instance.props,
+      'It looks like the %s component contains a line like this ' +
+        'in its constructor:\n\n' +
+        'this.state = props;\n\n' +
+        "This is not recommended because any further updates to props won't be " +
+        "reflected in the state. In most cases, you don't need to keep state and " +
+        'props in sync. Instead, use the props directly. If you need to calculate ' +
+        'something from the props, do it during the rendering. If you need to ' +
+        'share the state between several components, move it to their closest ' +
+        'common ancestor and pass it down as props to them.',
+      getComponentName(workInProgress) || 'Unknown',
```

Comment:
Just a small nit: it would probably be a better idea to pass `a component` instead of `Unknown` if no component name is found.

**Your example:**

```
It looks like the Unknown component contains a line ...
```

**Suggestion:**
```
A component contains a line ...
```

or if the component name exists

```
Foo contains a line like this...
```

This also would follow the consistency of handling missing component names in error messages. See: [ReactFiberTreeReflection.js](https://github.com/facebook/react/blob/master/packages/react-reconciler/src/ReactFiberTreeReflection.js#L77), [ReactDOM.js](https://github.com/facebook/react/blob/master/packages/react-dom/src/client/ReactDOM.js#L1139)

Reply:
I'm inclined to agree, but as I searched through the code base I saw the `getComponentName(*) || 'Unknown'` pattern used 22 times while `Component` or `A component` were only used 7 and 8 times respectively. In fact, the most recent commits by @gaearon, @acdlite, and @bvaughn all use `Unknown` (#12028, #12359, #11455). I'll let one of them chime in about whether they think this should be changed.

Reply:
It's true that "Unknown" is probably a more common fallback name, but it's not strictly necessary to use that pattern. That being said, I think this warning is a bit verbose and maybe tries to cover too broad a topic (e.g. sharing state between several components).

At first glance, I would prefer a shorter message, something like:
```js
warning(
  instance.state !== instance.props,
  "%: It is not recommended to assign props directly to state. " +
    "Updates to props won't be reflected in state. " +
    "It's usually better to just use props directly.",
  getComponentName(workInProgress) || "Unknown"
);
```